frontend/src/tests/unit/validators.test.js

```diff
@@ -0,0 +1,221 @@
+    // given
+    const examplePassword = ''
+    const exampleConfirmPassword = 'password'
+
+    // when
+    const validatorMessage1 = validateConfirmPassword(examplePassword, exampleConfirmPassword)
+    const validatorMessage2 = validateConfirmPassword(exampleConfirmPassword, examplePassword)
+
+    // then
+    expect(validatorMessage1).toBe(FIELD_REQUIRED)
+    expect(validatorMessage2).toBe(FIELD_REQUIRED)
+  })
+  it('should return DIFFERENT_PASSWORDS error when passwords are different', () => {
+    // given
+    const examplePassword = 'Psw1'
+    const exampleConfirmPassword = 'Psw2'
+
+    // when
+    const validatorMessage1 = validateConfirmPassword(examplePassword, exampleConfirmPassword)
+    const validatorMessage2 = validateConfirmPassword(exampleConfirmPassword, examplePassword)
+
+    // then
+    expect(validatorMessage1).toBe(DIFFERENT_PASSWORDS)
+    expect(validatorMessage2).toBe(DIFFERENT_PASSWORDS)
+  })
+})
+
+describe('Email validator tests:', () => {
+  it('should return FIELD_REQUIRED error if email string is empty', () => {
+    // given
+    const exampleEmail = ''
+
+    // when
+    const validatorMessage1 = validateEmail(exampleEmail, AccountType.STUDENT)
+    const validatorMessage2 = validateEmail(exampleEmail, AccountType.PROFESSOR)
+
+    // then
+    expect(validatorMessage1).toBe(FIELD_REQUIRED)
+    expect(validatorMessage2).toBe(FIELD_REQUIRED)
+  })
+  it('should return INCORRECT_EMAIL error for bad email string', () => {
+    // given
+    const exampleEmail1 = 'testEmail'
+    const exampleEmail2 = 'testEmail@'
+    const exampleEmail3 = 'testEmail@test.'
+    const exampleEmail4 = 'test@test.com'
+    const exampleEmail5 = 'test@test.pl'
+    const exampleEmail6 = 'test@agh.edu.pl.test.pl'
+    const exampleEmail7 = '@agh.edu.pl'
+    const exampleEmail8 = '123456@agh.edu.pl'
+
+    // when
+    const validatorMessage1 = validateEmail(exampleEmail1, AccountType.STUDENT)
+    const validatorMessage2 = validateEmail(exampleEmail2, AccountType.PROFESSOR)
+    const validatorMessage3 = validateEmail(exampleEmail3, AccountType.PROFESSOR)
+    const validatorMessage4 = validateEmail(exampleEmail4, AccountType.STUDENT)
+    const validatorMessage5 = validateEmail(exampleEmail5, AccountType.PROFESSOR)
+    const validatorMessage6 = validateEmail(exampleEmail6, AccountType.STUDENT)
+    const validatorMessage7 = validateEmail(exampleEmail7, AccountType.PROFESSOR)
+    const validatorMessage8 = validateEmail(exampleEmail8, AccountType.STUDENT)
+
+    // then
+    expect(validatorMessage1).toBe(INCORRECT_EMAIL)
+    expect(validatorMessage2).toBe(INCORRECT_EMAIL)
+    expect(validatorMessage3).toBe(INCORRECT_EMAIL)
+    expect(validatorMessage4).toBe(INCORRECT_EMAIL)
+    expect(validatorMessage5).toBe(INCORRECT_EMAIL)
+    expect(validatorMessage6).toBe(INCORRECT_EMAIL)
+    expect(validatorMessage7).toBe(INCORRECT_EMAIL)
+    expect(validatorMessage8).toBe(INCORRECT_EMAIL)
```

Comment:
Okej, to trzeba zrobić refactor - zamiast rozpisywać każdy pojedynczy case tak, można wykorzystać `it.each` po rozbiciu na przypadki dla studenta i prowadzącego do zmiennej z tablicą stringów. Wtedy caaały ten test zwinie się do:

```ts
it.each(invalidEmailsStudent)(
    'should return INCORRECT_EMAIL error for bad email string for student email %s',
    (email) => {
      // when
      const validatorMessage = validateEmail(email, AccountType.STUDENT)

      // then
      expect(validatorMessage).toBe(INCORRECT_EMAIL)
    }
  )
```

(no i tablicy stringów na górze plików.)

Wtedy nawet mamy to rozbite na pojedyncze przypadki w konsoli, więc w razie czego można łatwo zobaczyć na czym się sypnęło:



W zasadzie to nie jest chyba jedyny test który uda się tak zwinąć, i myślę że warto - wtedy dodanie case do dowolnego testu to kwestia dopisania stringa do tablicy zamiast trzech linijek





![Code_eCkItwq95s](https://user-images.githubusercontent.com/56168780/194646288-7f6b4d85-dccd-49f0-9451-571d12623c65.png)

Reply:
Dzięki, nie znałem tego it.each :D
Pora się bardziej zagłębić w dokumentację jesta 😅